### PR TITLE
Add some ttir.gather -> ttir.embedding cases

### PR DESCRIPTION
**Pull request by sdjukicTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-mlir/issues/1350)
(Not all cases are covered)

### Problem description
`ttir.gather` that is lowered to `ttir.embedding` in TTIR to TTIR decomposition was implemented with constraints that can be loosened up, with more cases covered.

### What's changed
-`ttir.embedding` supports indexing the first dimension of a 2D tensor, and getting slices of full size from the second dimension. We call this tensor `weight`.
-The other operand is a tensor that has the start indices. This tensor can be 1D or 2D and we call it `input`.
-Output shape is the shape of input with the last dimension of weights appended (indices are replaced with corresponding slices from the weights).

Cases of `ttir.gather` that will be supported after this PR are those where:
-The dimensions we are indexing have slice size 1 and don't appear in the output as offset dimensions -> collapsedSliceDims are the same as startIndexMap
-For all other dimensions, slice size has to be equal to dimension size (because embedding takes the whole row) -> sliceSizes = [1, 1, ...fullDim, fullDim, ...]
=> This mean we can permute and reshape the operand appropriately. First we permute so all of the indexing dims are at the start of the operand. We then take the first few dimensions that we are indexing and flatten them together to be the first dimension. We also flatten the other dimensions together to be the second dimension.
This 2D tensor is `weight` for `ttir.embedding`.

-There are no input or start indices batching dimensions, as this can't be lowered directly to `ttir.embedding`.

-start indices tensor can't be 1D  when we are indexing multiple dimensions. This is because `ttir.matmul` that transforms start indices needs a 2D tensor.
=> This means we can reshape start indices tensor appropriately. If it is already rank 1 or 2, we leave it as to not add unnecessary reshapes. Otherwise we flatten the dimensions together.
=> If more than one dimension of operand is being indexed, we add `ttir.matmul` to transform the indices to match the flattened operand. This requires a permutation if index vector dim isn't the last dimension.
This 1D or 2D tensor is `input` for `ttir.embedding`.

-Offset dimensions can be any length. Output shape of 'ttir.embedding' It will always have the slices as last dimension (and offset dimensions represent where the slices are), so we need a permute op if offset dimensions aren't the last dimensions of the output.
=> If necessary, output will be reshaped as well, to match `ttir.gather` output. This will be the case when we need to restore dimensions that were flattened.

### Notes
Will add more tests and negative tests.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by azecevicTT** [comment]
Thanks for the very detailed PR description! I want to point out a few nits regarding terminology, since this is going to end up as the PR message:

>-ttir.embedding supports indexing the first dimension of a 2D tensor, and getting slices of full size from the second dimension. We call this tensor weights tensor or input. I agree that `input` is not the best name choice, but it is what it is for now.

To be precise, we model `ttir.embedding` with `input` and `weight` parameters. In that terminology `weight` is the tensor you've described, and `input` is an indexing tensor.

>-Output shape is the shape of start indices with the last dimension of the input appended (indices are replaced with corresponding slices from the input).

Let's stick with `ttir.embedding` terminology when we talk about it, so `start_indices` is `input` and `input` is `weight`.

>-We index the input tensor at dimension 0 -> startIndexMap = [0]

At least the good thing about naming is that the names are disjoint in `gather` and `embedding`, so let's use `operand` when talking about input in `gather` to disambiguate it.


I will continue review and try to give more meaningful insights once I get more familiar with the details of `gather` (there are a lot of them!). In the meantime I will leave some technical/stylistic comments.

**Review by sdjordjevicTT — APPROVED** [review]
Thanks for the detailed description, comments, and tests @sdjukicTT, this looks good to me, maybe @azecevicTT  can take one more look when he is back in the office next week.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 1]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+    auto numIndexingDims = op.getStartIndexMap().size();
+    auto indexVectorDim = op.getIndexVectorDim();
+
+    llvm::SmallVector<int64_t> startIndicesPermutation;
+    for (size_t i = 0; i < startIndicesType.getShape().size(); i++) {
+      if (static_cast<int>(i) != indexVectorDim) {
+        startIndicesPermutation.push_back(i);
+      }
+    }
+    startIndicesPermutation.push_back(indexVectorDim);
```
Comment (azecevicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:663]:
```suggestion
    llvm::SmallVector<int64_t> startIndicesPermutation = llvm::filter_to_vector(
        llvm::seq<int64_t>(startIndicesType.getRank()),
        [&indexVectorDim](int64_t idx) { return idx != indexVectorDim; });
    startIndicesPermutation.push_back(indexVectorDim);
```

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 2]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+  // If we are indexing multiple dimes of input, we need to adjust start
+  // indices to represent indices that index one flattened dimension.
+  // - indexVectorDim represents in what dimension are indices, so first we
+  // permute to make sure it is the last dimension
+  // - matmul doesn't work with integers (which startIndices are when lowered
+  // form SHLO), so a typecast is added
+  // - then we add matmul to transform the indices
+  // Example: indexingDimsSizes = [3, 5], startIndices[...] = (i, j) ->
+  // startIndices[...] = 5 * i + j (because reshaped indexingDimSize is 15)
+  ttir::MatmulOp transformStartIndices(ConversionPatternRewriter &rewriter,
```
Comment (azecevicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:649]: All of these helper functions can be static.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 3]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+            rewriter, op.getLoc(), permutedStartIndicesShape,
+            startIndicesType.getElementType(), startIndicesType.getEncoding(),
+            startIndices, startIndicesPermutation)
+            .getResult();
+
+    // Typecast op because matmul needs float operands.
+    auto typecastResultType = startIndicesPermuted.getType().clone(
+        mlir::Float32Type::get(getContext()));
+    ttir::TypecastOp typecastOp = ttir::utils::createDPSOp<ttir::TypecastOp>(
+        rewriter, op->getLoc(), typecastResultType, startIndicesPermuted);
```
Comment (azecevicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:678]: This is a mathematically very sound idea, but I wonder if it will work correctly with the FP values because of numeric stability. If `fp -> integer` typecast rounds towards zero and some `integer -> fp` conversion ends up being little bit less that actual value then the result will be incorrect. For example (it's contrieved example, but it can actually happen for some other numbers):
```
typecast: 15 (i) -> 14.9 (fp)
mul: 14.9 * 3.0 -> 44.7
typecast: 44.7 (fp) -> 44 (i)
```

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 4]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+
+    // Const op with correct weights to matmul indices with.
+    llvm::SmallVector<float> indexWeight(numIndexingDims);
```
Comment (azecevicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:681]: I think the accepted term is `strides`.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 5]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+    auto matmulResultType = mlir::RankedTensorType::get(
+        matmulResultShape, Float32Type::get(getContext()));
+
+    return ttir::utils::createDPSOp<ttir::MatmulOp>(
+        rewriter, op->getLoc(), matmulResultType, typecastOp.getResult(0),
+        constantOp);
+  }
+
+  // Helper that reshapes start indices to reduce number of dims as it can't be
+  // 3 or greater.
```
Comment (azecevicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:707]: It's unclear where does 3 comes from here.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 6]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+  ttir::ReshapeOp reshapeStartIndices(
+      ConversionPatternRewriter &rewriter, Location loc,
+      ::mlir::TypedValue<::mlir::RankedTensorType> startIndices) const {
+    auto startIndicesType = startIndices.getType();
+    auto startIndicesShape = startIndicesType.getShape();
+    llvm::SmallVector<int64_t, 1> newStartIndicesShape = {1};
+    for (size_t i = 0; i < startIndicesShape.size(); i++) {
+      newStartIndicesShape[0] *= startIndicesShape[i];
+    }
```
Comment (azecevicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:716]:
```suggestion
    auto startIndicesShape = startIndices.getType().getShape();
    llvm::SmallVector<int64_t, 1> newStartIndicesShape {
      std::accumulate(startIndicesShape.begin(), startIndicesShape.end(), int64_t{1}, std::multiplies<>())
    };
```

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 7]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+
+  // In output, dims other than offsetDims map to startIndices shape, and
+  // offsetDims map to input slices. After ttir.embedding all offseDims are
+  // flattened to the last dim of output. First we reshape that output to
+  // recover lost dims, then we permute them so offset dims are where the
+  // attribute states.
+  // Example: expectedOutputShape = [2, 3, 4, 5], offsetDims = [1, 3]
+  // -> embeddingOutputShape = [2, 4, 15] -reshape-> [2, 4, 3, 5] -permute-> [2,
+  // 3, 4, 5]
+  ttir::PermuteOp reshapeAndPermuteOutput(
```
Comment (azecevicTT) [inline lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:728]: I get that these are helper functions so they don't have to be the most general one, but is it possible to parameterize these function in such way that we have
```cpp
reshape(input, targetShape)
permute(input, targetPermutation)
```
and then just compose these functions (directly in the main function, or in some other helper function). It will make it easier for maintenance.

**Review by azecevicTT — COMMENTED** [review]
Reviewed mainly helper functions for now, overall it looks good.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 8]
```diff
@@ -519,69 +517,180 @@ struct GatherToEmbeddingConversionPattern
...
+          reshapeStartIndices(rewriter, op->getLoc(), startIndicesTransformed);
     }
 
-    // convert gather to embedding, use reshaped input if needed
-    ttir::EmbeddingOp embeddingOp = rewriter.create<ttir::EmbeddingOp>(
-        op.getLoc(), op.getResult().getType(), input, op.getOperands()[0],
-        op.getOutput());
+    // Calculate a new shape for output: this is new start indices shape + last
+    // dim of input shape.
+    std::vector<int64_t> newOutputShape = startIndices.getType().getShape();
```
[the remark this answers is not in the document]
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: Changed

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 9]
```diff
@@ -519,69 +517,180 @@ struct GatherToEmbeddingConversionPattern
...
+            .getResult();
+
+    // Typecast op because matmul needs float operands.
+    auto typecastResultType = startIndicesPermuted.getType().clone(
+        mlir::Float32Type::get(getContext()));
+    ttir::TypecastOp typecastOp = ttir::utils::createDPSOp<ttir::TypecastOp>(
+        rewriter, op->getLoc(), typecastResultType, startIndicesPermuted);
+
+    // Const op with correct weights to matmul indices with.
+    std::vector<float> indexWeight(numIndexingDims);
```
[the remark this answers is not in the document]
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: Changed

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 10]
```diff
@@ -519,69 +517,180 @@ struct GatherToEmbeddingConversionPattern
...
 
+    rewriter.replaceOp(op, reshapeAndPermuteOutput(rewriter, op->getLoc(),
+                                                   embeddingOp, op.getOutput(),
+                                                   op.getOffsetDims()));
     return success();
   }
 
 private:
+  ttir::PermuteOp
+  permuteInput(ConversionPatternRewriter &rewriter, Location loc,
```
[the remark this answers is not in the document]
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:599]: Added a comment with an example

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 11]
```diff
@@ -519,69 +517,180 @@ struct GatherToEmbeddingConversionPattern
...
+    for (size_t i = 0; i < numIndexingDims; i++) {
+      newInputShape[0] *= inputShape[i];
+    }
+    for (size_t i = numIndexingDims; i < inputShape.size(); i++) {
+      newInputShape[1] *= inputShape[i];
+    }
+    return createReshapeOp(rewriter, loc, input, newInputShape);
+  }
+
+  ttir::MatmulOp transformStartIndices(ConversionPatternRewriter &rewriter,
```
[the remark this answers is not in the document]
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: Added comment

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 12]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
-      // check if sliceSizes is [1, hiddenDim]
-      if (sliceSizes[0] != 1 || sliceSizes[1] != hiddenDim) {
-        return rewriter.notifyMatchFailure(op, "Did not satisfy sliceSizes");
-      }
+    // Check if start indices tensor isn't 1D when we are indexing multiple
+    // dimensions because of matmul restrictions.
+    if (startIndexMap.size() > 1 && startIndicesShape.size() == 1) {
+      return rewriter.notifyMatchFailure(
+          op,
+          "Did not satisfy indexVectorDim != 0 when startIndexMap.size() > 1");
```
[the remark this answers is not in the document]
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: Fixed

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp [hunk 13]
```diff
@@ -444,144 +444,321 @@ struct GatherToEmbeddingConversionPattern
...
+    for (size_t inputShapeIndex = 0;
+         inputShapeIndex < inputType.getShape().size(); inputShapeIndex++) {
+      if (startIndexMapIndex < startIndexMap.size() &&
+          startIndexMap[startIndexMapIndex] ==
+              static_cast<long>(inputShapeIndex)) {
+        startIndexMapIndex++;
+        continue;
+      }
+      inputPermutation.push_back(inputShapeIndex);
+    }
```
[the remark this answers is not in the document]
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: I agree, changed

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 4 above)
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: Changed

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 7 above)
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: Because these functions are short and have a comment that explains how we choose target shape and permutation with an example, I think it's okay for maintenance, do you agree?

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 3 above)
Comment (sdjukicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:672]: It is the only way I thought of. Maybe I should leave it but add a warning?

Comment (azecevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp:672]: Yeah, warning sounds like a good idea if someone hit a problem related to it.

File: lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp (hunk 7 above)
Comment (azecevicTT) [reply lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp]: It's okay for now, if we decide to support even more cases in the future we should probably refactor it at some point.

**Review by azecevicTT — APPROVED** [review]
LGTM, thanks for adding a lot of test cases!
